src/main/java/de/tum/in/www1/artemis/service/FileUploadSubmissionService.java

```diff
@@ -0,0 +1,325 @@
+package de.tum.in.www1.artemis.service;
+
+import java.security.Principal;
+import java.time.ZonedDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.messaging.simp.SimpMessageSendingOperations;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.server.ResponseStatusException;
+
+import de.tum.in.www1.artemis.domain.*;
+import de.tum.in.www1.artemis.domain.enumeration.*;
+import de.tum.in.www1.artemis.repository.*;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+@Service
+@Transactional
+public class FileUploadSubmissionService extends SubmissionService {
+
+    private final Logger log = LoggerFactory.getLogger(FileUploadSubmissionService.class);
+
+    private final FileUploadSubmissionRepository fileUploadSubmissionRepository;
+
+    private final ResultService resultService;
+
+    private final ResultRepository resultRepository;
+
+    private final ParticipationService participationService;
+
+    private final StudentParticipationRepository studentParticipationRepository;
+
+    private final SimpMessageSendingOperations messagingTemplate;
+
+    public FileUploadSubmissionService(FileUploadSubmissionRepository fileUploadSubmissionRepository, SubmissionRepository submissionRepository, ResultRepository resultRepository,
+            ParticipationService participationService, UserService userService, StudentParticipationRepository studentParticipationRepository,
+            SimpMessageSendingOperations messagingTemplate, ResultService resultService) {
+        super(submissionRepository, userService);
+        this.fileUploadSubmissionRepository = fileUploadSubmissionRepository;
+        this.resultRepository = resultRepository;
+        this.participationService = participationService;
+        this.studentParticipationRepository = studentParticipationRepository;
+        this.messagingTemplate = messagingTemplate;
+        this.resultService = resultService;
+    }
+
+    /**
+     * Handles file upload submissions sent from the client and saves them in the database.
+     *
+     * @param fileUploadSubmission the file upload submission that should be saved
+     * @param fileUploadExercise   the corresponding file upload exercise
+     * @param principal            the user principal
+     * @return the saved file upload submission
+     */
+    @Transactional
+    public FileUploadSubmission handleFileUploadSubmission(FileUploadSubmission fileUploadSubmission, FileUploadExercise fileUploadExercise, Principal principal) {
+        Optional<StudentParticipation> optionalParticipation = participationService.findOneByExerciseIdAndStudentLoginAnyState(fileUploadExercise.getId(), principal.getName());
```

Comment:
This again is a copy & pasted method from the `TextSubmissionService`. Please find a way to streamline this - we can't just duplicate the code for every new exercise type.

There are more examples in this file but I won't list every one of it here (you can check this yourself in e.g. IDEA Ultimate as duplicated code is marked in the editor).

Reply:
As discussed we refactor this in a later PR